compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -111,18 +85,23 @@ object TypeTestsCasts {
       // conforms to the type skeleton pre.F[_]. Then it goes on to check
       // if P1 <: P, which means the type arguments in P are trivial,
       // thus no runtime checks are needed for them.
-      P1 <:< X
+      withMode(Mode.GadtConstraintInference) {
+        TypeComparer.constrainPatternType(P1, X, widenParams = false)
+        debug.println(TypeComparer.explained(_.constrainPatternType(P1, X, widenParams = false)))
```

Comment:
I think I understand the intent here, but it's a bit surprising that we don't widen the params, given that doing so is necessary for soundness. What would be a motivating example for _not_ widening them here?

Reply:
The motivation for not widening is something specific to the algorithm: here for `P1 = pre.F[Xs]`, all `Xs` are type variables.

Reply:
I feel very happy the work on the GADT makes the checking so elegant that we remove a lot of ad-hoc tweaks. The fact that the algorithm is simpler and it found more bugs in the compiler gives more confidence.

Reply:
One example is the following:

```Scala
  // tests/neg-custom-args/isInstanceOf/JavaSeqLiteral.scala
  trait Tree[-T]

  class JavaSeqLiteral[-T] extends Tree[T]

  trait Type

  class DummyTree extends JavaSeqLiteral[Any]

  def foo1(tree: Tree[Type]) =
    tree.isInstanceOf[JavaSeqLiteral[Type]]
```

In checking whether the test `tree.isInstanceOf[JavaSeqLiteral[Type]]` is realizable, we want to constrain `JavaSeqLiteral[X] <: Tree[Type]`, such that we can infer `X = Type` and `JavaSeqLiteral[X] <:< JavaSeqLiteral[Type]`.

If we perform widening, we will get `X = Nothing`, and we don't have `JavaSeqLiteral[X] <:< JavaSeqLiteral[Type]` any more.

Reply:
Ok, I think what we're doing here is sound, but with small changes we could easily make it unsound. I'll need to add a comment explaining what's going on, and then we can merge the PR.

Reply:
Yes, I'll add a comment explaining the code.

Reply:
Comment added in aa296ba